### ISSUE TITLE
Bug in nextmap plugin: mp_chattime gets increased on map change

### ISSUE THREAD

**Issue by rtxa** [alliedmodders/amxmodx#855]
## Issue
There's an issue with the nextmap plugin, when intermission is triggered and you do a map change, e.g. using amx_map, mp_chattime gets increased by 2.

When this is done several times, the mp_chattime will have reach a high value, making the time for a map change very long.

The cause of this is found in the function event `changeMap()`, most specifically in this line:
https://github.com/alliedmodders/amxmodx/blob/2559fcf00aecb19fb6e6f8de846ae37cbcd6d572/plugins/nextmap.sma#L112

The function `delayedChange()` never gets executed because it's called by a task with the value of `[mp_chattime original value + two seconds]`, and if you do the map change before of that, the bug is triggered.

https://github.com/alliedmodders/amxmodx/blob/2559fcf00aecb19fb6e6f8de846ae37cbcd6d572/plugins/nextmap.sma#L95-L101

## Solution
One solution that has come to my mind is to the reset mp_chattime to his original value in plugin_end()

```pawn
new g_changeMapCalled;

public delayedChange(param[])
{
	engine_changelevel(param)
}

public plugin_end()
{
	if (g_mp_chattime && g_changeMapCalled) {
		set_pcvar_float(g_mp_chattime, get_pcvar_float(g_mp_chattime) - 2.0)
	}
}

public changeMap()
{
	new string[32]
	new Float:chattime = g_mp_chattime ? get_pcvar_float(g_mp_chattime) : 10.0;	// mp_chattime defaults to 10 in other mods
	
	if (g_mp_chattime) {
		set_pcvar_float(g_mp_chattime, chattime + 2.0)		// make sure mp_chattime is long
		g_changeMapCalled = true;
	}
	new len = getNextMapName(string, charsmax(string)) + 1
	set_task(chattime, "delayedChange", 0, string, len)	// change with 1.5 sec. delay
}
```

This way, you will make sure the mp_chattime is reseted always to his original value, making it more reliable.

Please, tell me what you think, I tested the solution for about a week and I didn't see the issue coming back since then.
If you have a better solution or an improvement feel free to post it, thanks.

**Comment by stefanofabi** [comment]
executing the mp_chattime command with rcon returns the value incremented by 2?

**Comment by rtxa** [comment]
> 
> 
> executing the mp_chattime command with rcon returns the value incremented by 2?

When final score (intermission) is displayed and you use `amx_map`, after map change (you will notice that the map changes so quickly), checking the value with `amx_cvar mp_chattime`, will show you that the value is incremented by 2.

**Comment by stefanofabi** [comment]
I don't understand the logic of why not allow mp_chattime 0.

I also do not understand why it adds 2 to the time, I see it illogical.

I have removed these lines and the error you are commenting on goes perfectly.


`public delayedChange(param[])
{
	engine_changelevel(param)
}

public changeMap()
{
	new string[32]
	new Float:chattime = get_pcvar_float(g_mp_chattime);	// mp_chattime defaults to 10 in other mods
	
	new len = getNextMapName(string, charsmax(string)) + 1
	set_task(chattime, "delayedChange", 0, string, len)	// change with 1.5 sec. delay
}`

**Comment by rtxa** [comment]
> I also do not understand why it adds 2 to the time, I see it illogical.
> 
> I have removed these lines and the error you are commenting on goes perfectly.

I don't understand, are you saying that you can't reproduce the bug? 

> Edit: By default HLDS sets mp_chattime > 0

Yeah, HL sets mp_chattime in server startup to 10 by default, but if you have in your server.cfg mp_chattime, you are not going to be able to reproduce the bug, because server.cfg gets executed on every map change, remove it in case you have it.

PD: I have updated the code, the oldest one didn't solve it the bug completly.

**Comment by stefanofabi** [comment]
What is executed at each map change is amxx.cfg.
The server.cfg file only runs at server startup and then never again.

The error if I could do it, but as I was saying, change those lines of code, then compile the plugin again and the error was eliminated.

**Comment by rtxa** [comment]
> 
> 
> What is executed at each map change is amxx.cfg.
> The server.cfg file only runs at server startup and then never again.
> 
> The error if I could do it, but as I was saying, change those lines of code, then compile the plugin again and the error was eliminated.
> 
> Sorry for my very bad English.

Yes, server.cfg file only runs one time. Now I understand, you tested the solution and now the issue is fixed it, cool.

**Comment by fysiks1** [comment]
FYI, there is a cvar ("mapchangecfgfile") that sets a file to be executed on each map change.  So, if you set "mapchangecfgfile" to "server.cfg" then server.cfg will execute on each map change.  Setting it to some other file will allow you to have server.cfg as the startup settings and only the other file will be executed on map change.

In the case the workaround for this issue, you can either set "server.cfg" as the "mapchangecfgfile" or put `mp_chattime 10` in the file that is configured as the mapchangecfgfile.

**Comment by rtxa** [comment]
@fysiks1 Yes, that's a good workaround to reset it every map change. Still I would like this to get fixed, I'm doing a PR now...

**Comment by rtxa** [comment]
Issue has been fixed and merged with #1044 , closing issue...
